### PR TITLE
SWC-6742 - NoSearchResults on FavoritesPage

### PR DESCRIPTION
**Pull request by nickgros**
- Reworked automatic facet filters with TanStack Table to rely on existing APIs and remove custom API
- Updated FavoritesPage to use TanStack Table, added automatic sorting and filtering
- Moved NoSearchResults content to S3 because we were not re-theming the SVG

### CODE REVIEW

File: packages/synapse-react-client/src/assets/icons/NoSearchResults.tsx [hunk 1]
```diff
@@ -2,56 +2,13 @@ import { Box, BoxProps } from '@mui/material'
...
-        d="M273.24 69.8943C276.502 69.8943 279.146 67.25 279.146 63.9881C279.146 60.7262 276.502 58.0819 273.24 58.0819C269.978 58.0819 267.333 60.7262 267.333 63.9881C267.333 67.25 269.978 69.8943 273.24 69.8943Z"
-        fill="white"
-      />
-    </svg>
-  </Box>
+  <Box
+    {...props}
+    component={'img'}
+    src={
+      'https://s3.us-east-1.amazonaws.com/static.synapse.org/images/NoSearchResults.svg'
```
Comment (nickgros) [inline packages/synapse-react-client/src/assets/icons/NoSearchResults.tsx:9]: I put this file in S3 to simplify getting it in SWC. We aren't re-theming it so there's no reason for the SVG data to live in a React component.

File: packages/synapse-react-client/src/components/TanStackTable/ColumnHeaderEnumFilter.test.tsx [hunk 2]
```diff
@@ -14,52 +17,89 @@ type TestTableRow = {
   multiSelectWithCounts: string
   singleSelect: string
   numeric: number
+  defaultClientSideFacetCalculation: string
 }
 
 const columnHelper = createColumnHelper<TestTableRow>()
 
+function getFacetedUniqueValues(table: Table<TestTableRow>, columnId: string) {
```
Comment (nickgros) [inline packages/synapse-react-client/src/components/TanStackTable/ColumnHeaderEnumFilter.test.tsx:25]: Update test for component that creates facet filter UI using TanStack table data

File: packages/synapse-react-client/src/components/TanStackTable/ColumnHeaderEnumFilter.test.tsx [hunk 3]
```diff
@@ -14,52 +17,89 @@ type TestTableRow = {
...
+  if (columnId === 'multiSelect') {
+    return () =>
+      new Map([
+        ['foo', 0],
+        ['bar', 0],
```
Comment (nickgros) [inline packages/synapse-react-client/src/components/TanStackTable/ColumnHeaderEnumFilter.test.tsx:30]: TanStack table does not provide an API for facets without counts, so I added some bespoke logic to deal with this when all of the counts are 0. The TanStack table types do not allow this value to be null.

File: packages/synapse-react-client/src/components/TanStackTable/ColumnHeaderEnumFilter.tsx [hunk 4]
```diff
@@ -62,10 +82,6 @@ export function ColumnHeaderEnumFilter<TData = unknown, TValue = unknown>(
   const { column, title } = props
 
   const canMultiSelect = column.columnDef.meta?.enableMultipleSelect || false
-  const enumValues: ColumnFilterEnumValue<TValue>[] = useMemo(
-    () => column.columnDef.meta?.enumValues || [],
-    [column.columnDef.meta?.enumValues],
-  )
```
Comment (nickgros) [inline packages/synapse-react-client/src/components/TanStackTable/ColumnHeaderEnumFilter.tsx:68]: No longer use our custom `meta.enumValues` API

File: packages/synapse-react-client/src/components/TanStackTable/ColumnHeaderEnumFilter.tsx [hunk 5]
```diff
@@ -48,6 +46,28 @@ type ColumnHeaderEnumFilterProps<TData = unknown, TValue = unknown> = {
   title: string
 }
 
+function getMaybeFacetedUniqueValues<TData = unknown, TValue = unknown>(
+  column: Column<TData, TValue>,
+): {
+  value: TValue
+  count?: number
+}[] {
+  if (column.getFacetedUniqueValues()) {
```
Comment (nickgros) [inline packages/synapse-react-client/src/components/TanStackTable/ColumnHeaderEnumFilter.tsx:55]: Use `column.getFacetedUniqueValues` to get the faceted values

File: packages/synapse-react-client/src/components/TanStackTable/ColumnHeaderEnumFilter.tsx [hunk 6]
```diff
@@ -76,16 +92,22 @@ export function ColumnHeaderEnumFilter<TData = unknown, TValue = unknown>(
...
   const transformedFilterValues: RenderedFacetValue<TValue>[] = useMemo(
     () =>
-      enumValues.map(
-        (v: ColumnFilterEnumValue<TValue>): RenderedFacetValue<TValue> => {
+      getMaybeFacetedUniqueValues(column).map(
+        (facetValue): RenderedFacetValue<TValue> => {
+          let displayText: string = facetValue.value as unknown as string
+          if (column.columnDef?.meta?.getDisplayText) {
+            displayText = column.columnDef.meta.getDisplayText(facetValue.value)
+          }
```
Comment (nickgros) [inline packages/synapse-react-client/src/components/TanStackTable/ColumnHeaderEnumFilter.tsx:100]: Get the display text (which was previously included in the `enumValues` data) using a new custom function.

File: packages/synapse-react-client/src/components/dataaccess/UseAccessRequirementTable.tsx [hunk 7]
```diff
@@ -307,6 +293,20 @@ export function useAccessRequirementTable(
...
       },
+      getFacetedUniqueValues: (
+        table: Table<AccessRequirementSearchResult>,
+        columnId: string,
+      ) => {
+        if (columnId === 'type') {
+          // Hard-code the set of possible access requirement types with count = 0 to indicate that we don't have facet counts.
+          return () =>
+            new Map(ALL_ACCESS_REQUIREMENT_CONCRETE_TYPES.map(v => [v, 0]))
+        }
```
Comment (nickgros) [inline packages/synapse-react-client/src/components/dataaccess/UseAccessRequirementTable.tsx:304]: We want to provide faceted filter options, where we know all of the facet values, but we don't have facet counts. In this case, we provide a custom implementation that returns all of the facet values with counts set to 0.

File: packages/synapse-react-client/src/components/favorites/FavoritesPage.tsx [hunk 8]
```diff
@@ -27,15 +35,84 @@ export type Sort = {
...
+  )
+
+  const favoritesTable = useReactTable<EntityHeader>({
+    data: favorites,
+    columns: columns,
+    getCoreRowModel: getCoreRowModel(),
+    getFilteredRowModel: getFilteredRowModel(),
+    getSortedRowModel: getSortedRowModel(),
+    getFacetedRowModel: getFacetedRowModel(),
+    getFacetedUniqueValues: getFacetedUniqueValues(),
```
Comment (nickgros) [inline packages/synapse-react-client/src/components/favorites/FavoritesPage.tsx:114]: Added ability to filter on entity type. This client-side filter implementation is provided by TanStack Table.

File: packages/synapse-react-client/src/components/favorites/FavoritesPage.tsx [hunk 9]
```diff
@@ -51,154 +128,49 @@ export default function FavoritesPage() {
...
-          placeholder="Favorite Name"
-          value={searchText}
-          onChange={event => {
-            setSearchText(event.target.value)
-          }}
-        />
-      </div>
-
-      {sortedData && sortedData.length > 0 && (
-        <div className="bootstrap-4-backport">
```
Comment (nickgros) [inline packages/synapse-react-client/src/components/favorites/FavoritesPage.tsx:136]: Goodbye one more BS4 component :wave:

File: packages/synapse-react-client/src/components/favorites/FavoritesPage.tsx [hunk 10]
```diff
@@ -51,154 +128,49 @@ export default function FavoritesPage() {
...
+      <StyledTanStackTable table={favoritesTable} />
+      {!hasFavoritesMatchingQuery && !isLoading && (
+        <Stack sx={{ textAlign: 'center' }} my={2} gap={1}>
+          <NoSearchResults height={'150px'} />
+          {data?.results.length == 0 ? (
+            <p>You currently have no favorites</p>
+          ) : (
+            <p>No matching favorites found</p>
+          )}
+        </Stack>
```
Comment (nickgros) [inline packages/synapse-react-client/src/components/favorites/FavoritesPage.tsx:170]: Add the NoSearchResults UI

File: packages/synapse-react-client/src/types/tanstack-table-augmentation.d.ts [hunk 11]
```diff
@@ -23,9 +23,9 @@ declare module '@tanstack/react-table' {
 
     /** The type of filter that this column may have */
     filterVariant?: 'enumeration' // could be extended to include 'range', 'freeText', etc.
-    /** If the filterVariant is `enumeration`, this list captures the potential values that may be picked for filtering */
-    enumValues?: ColumnFilterEnumValue<TValue>[]
```
Comment (nickgros) [inline packages/synapse-react-client/src/types/tanstack-table-augmentation.d.ts:27]: Removed in favor of using one of the native `getFacetedUniqueValues` API. With this API, we can either 
* supply our own implementation that computes facet statistics (or retrieves them from a server)
* use the default implementation provided by TanStack Table that computes facet statistics client-side

File: packages/synapse-react-client/src/types/tanstack-table-augmentation.d.ts [hunk 12]
```diff
@@ -23,9 +23,9 @@ declare module '@tanstack/react-table' {
...
     /** If the filterVariant is `enumeration`, determines whether the user can filter by multiple values. */
     enableMultipleSelect?: boolean
+    /** Get the display text for a given value for this column, which may be used in non-cell UI such as the enumerated facet control. */
+    getDisplayText?: (value: TValue) => string
```
Comment (nickgros) [inline packages/synapse-react-client/src/types/tanstack-table-augmentation.d.ts:29]: the `enumValues` API provided a `displayName` field--we still need a way to pass this for the facet filter UI.

File: packages/synapse-react-client/src/components/TanStackTable/ColumnHeaderEnumFilter.test.tsx (hunk 3 above)
Comment (jay-hodgson) [reply packages/synapse-react-client/src/components/TanStackTable/ColumnHeaderEnumFilter.test.tsx:30]: It's a shame that the table type does not allow `undefined` or `null`

File: packages/synapse-react-client/src/components/favorites/FavoritesPage.tsx [hunk 13]
```diff
@@ -27,15 +35,84 @@ export type Sort = {
...
+        rel="noopener noreferrer"
+        href={`${PRODUCTION_ENDPOINT_CONFIG.PORTAL}Synapse:${ctx.row.original.id}`}
+      >
+        {ctx.row.original.name}
+      </a>
+    ),
+    enableGlobalFilter: true,
+    header: props => <ColumnHeader {...props} title={'Name'} />,
+    enableColumnFilter: false,
+    sortingFn: 'alphanumericCaseSensitive',
```
Comment (jay-hodgson) [inline packages/synapse-react-client/src/components/favorites/FavoritesPage.tsx:59]: Is this expected to be case sensitive?

**Review by jay-hodgson — APPROVED** [review]

File: packages/synapse-react-client/src/components/favorites/FavoritesPage.tsx (hunk 13 above)
Comment (nickgros) [reply packages/synapse-react-client/src/components/favorites/FavoritesPage.tsx:62]: No, probably not. Will change it

**Pull request merged**